posthog/models/kafka_engine_dlq/sql.py

```diff
@@ -0,0 +1,21 @@
+KAFKA_ENGINE_DLQ_BASE_SQL = """
+CREATE MATERIALIZED VIEW {database}.{kafka_table_name}_dlq
+(
+    `topic` String,
+    `partition` Int64,
+    `offset` Int64,
+    `raw` String,
+    `error` String
+)
+ENGINE = MergeTree
+ORDER BY (topic, partition, offset)
+SETTINGS index_granularity = 8192 AS
+SELECT
+    _topic AS topic,
+    _partition AS partition,
+    _offset AS offset,
+    _raw_message AS raw,
+    _error AS error
+FROM {database}.{kafka_table_name}
```

Comment:
Q: How does this work (do you have a link to documentation)?

Normally when you select from kafka tables the message would be committed and not available for subsequent selects, meaning an naive interprentation of this schema leads to ~50% of all messages getting lost.

Do materialized views work differently and are all triggered per block?

Reply:
So I started from [here](https://kb.altinity.com/altinity-kb-integrations/altinity-kb-kafka/error-handling/) and ended up reading source code [here](https://github.com/ClickHouse/ClickHouse/pull/21850). I also validated that having both materialized views running side by side does not cause any issues. My view of the world is that the Kafka engine works like this:

1. Poll in the background
2. Drop messages into a stream
3. A `SELECT` commits the offset of the messages pulled from the stream

Now, when you add the setting `kafka_handle_error_mode='stream'` for every errored message, it will drop two entries in the stream, one with all default values and `_error = ''` and the other with `_error` and `_raw_message`. Then, provided you have the correct `WHERE` in the MV (`WHERE length(_error) > 0`), all the error entries will go to one MV and the other messages will continue going to the other materialized view. i.e. the errors MV doesn't move the offset any further than it would have otherwise. Well, at least under normal conditions.

The one case I'm unsure about is if consuming the error entry moves forward the offset or not. If it doesn't, there's one potential edge case:

1. Messages 1, 2, 3, and 3_error get put in the stream
2. Default MV consumes message 1
3. Errors MV consumes message 3_error and commits the offset
4. ClickHouse server dies and the in-memory (?) stream is lost
5. Message 2 is never consumed when the server comes back up

I'll test this out by detaching an MV and killing the server locally. Side note that I'd have hoped this was considered in https://github.com/ClickHouse/ClickHouse/pull/21850#issue-834415807 as this behavior is what was recommended in the PR

Reply:
Hmm... somehow seems the stream is coordinated - detaching the main MV just halts the whole thing

Reply:
Thanks! This clears things up somewhat.

Thinking a bit further: When writing migrations should we now detach both of the MVs? Could this introduce issues where one is removed faster than the other?

Reply:
I'm trying to ask Altinity and ClickHouse about these things. Will also maybe spend some more time testing + reading source code